### ISSUE TITLE
LightGBM installs fine on Sierra, but doesn't seem to be running in parallel

### ISSUE THREAD

**Issue by ledmaster** [Microsoft/LightGBM#89]
Hello,

I installed LightGBM on my computer, which runs macOS Sierra, using gcc/g++ 6, but when I run it and monitor via htop, it doesn't use all the threads, even when I specify the number in the parameters. Running the same code in a cloud instance with Ubuntu 14.04 uses all the threads.

Another test I did was timing the script I am running. When I use XGB the "user" time is higher than "real", like in this example: 

**real	12m42.957s
user	31m41.828s**

But with LightGBM I get this:

**real	9m44.934s
user	9m57.032s**

I would like to ask if anybody had issues like this before, and if someone knows what can be the reason.

Thanks!

**Comment by guolinke** [comment]
@ledmaster can you provide the config/parameters ?

**Comment by ledmaster** [comment]
@guolinke Sure, here it is:

learning_rate=0.034
num_leaves=200
valid=/path_to/Xts_fold_2
bagging_freq=100
bagging_fraction=0.8
metric_freq=1
tree_learner=feature
feature_fraction=0.3
metric=l2
max_bin=255
early_stopping_round=0
application=regression
num_iterations=1000
bagging_seed=3
num_threads=4
min_data_in_leaf=8
feature_fraction_seed=2
data=/path_to/Xtr_fold_2
task=train
output_model=LightGBM_model.txt

I now know it will probably not make a difference, but I tried with tree_learner serial, feature and data. Same results.

Thanks!

**Comment by guolinke** [comment]
@chivee , can you help to test this on OS X?

**Comment by wxchan** [comment]
Here is my time information on Macbook Air (macOS Sierra, gcc5.3) with [experiments data](https://github.com/Microsoft/LightGBM/wiki/Experiments#comparison-experiment), num_threads=4, might help:

Higgs
real	39m48.687s
user	67m33.737s
sys	3m25.816s

MS LTR
real	22m55.961s
user	69m55.512s
sys	1m26.549s

**Comment by ledmaster** [comment]
I tried reinstalling, and compiling with gcc5.3, but it's still not working. Both gcc6 and gcc5.3 are installed without multilib.

Here is a small example I ran: 

real	1m17.821s
user	1m27.658s
sys	0m58.811s

**Comment by wxchan** [comment]
@ledmaster The reason might not be gcc. I have the same issue running 1-minute-job, but it seems normal with longer time jobs. I am not familiar with the reasons behind this, just provide some information.

**Comment by guolinke** [comment]
@ledmaster  you can try to run it with more iterations and check again?

**Comment by ledmaster** [comment]
@guolinke Sure, here are the results for ~20k iterations:

real	34m19.882s
user	39m39.402s
sys	21m27.047s

I tried upgrading cmake, Xcode, and reinstalled using gcc-6 again, but it seems this was not the issue.

**Comment by guolinke** [comment]
@ledmaster 
can you add following code after this line https://github.com/Microsoft/LightGBM/blob/master/src/application/application.cpp#L34:
```
int num_threads = 1;
#pragma omp parallel
#pragma omp master
{
  num_threads = omp_get_num_threads();
}
Log::Info("Number of threads : %d", num_threads);
omp_set_num_threads(4);
#pragma omp parallel
#pragma omp master
{
  num_threads = omp_get_num_threads();
}
Log::Info("Number of threads after explicit set to 4: %d", num_threads);
```
Then paste log to here?

**Comment by ledmaster** [comment]
@guolinke Sure, here it is: 

[LightGBM] [Info] Finished loading parameters
[LightGBM] [Info] Number of threads : 4
[LightGBM] [Info] Number of threads after explicit set to 4: 4
[LightGBM] [Info] Finished loading data in 0.648777 seconds
[LightGBM] [Info] Number of data: 150654, number of features: 136
[LightGBM] [Info] Finished initializing training
[LightGBM] [Info] Started training...

With this code, the time relation between real and user improves a bit:

real	35m36.783s
user	42m53.790s
sys	23m7.899s

htop screenshot while running:

![captura de tela 2016-11-22 as 13 07 24](https://cloud.githubusercontent.com/assets/9610024/20530836/d394d3d8-b0bb-11e6-9688-9132a68a528d.png)

Thank you for the time you guys are dedicating to this issue. LightGBM is really an amazing tool.

**Comment by guolinke** [comment]
I think the reason is:
1. It only has 2 cpu cores in your computer, the best setting is set num_threads=2(though it has 4 threads).
2. The test data is small, the computation cost is small, but the cost of open/close threads remain same. So it doesn't have much benefit with multi-threading. You can try a bigger data and try again.

Thanks!

**Comment by ledmaster** [comment]
@guolinke I see, I will do this test.

Just for some additional information that may be useful. The same code/dataset that runs "single threaded" in my laptop (4 threads / 2 cores) has this time difference with an Ubuntu 14.04 machine with 8 vCPUs (4 cores):

real    9m49.343s
user    47m27.944s
sys     0m29.788s

The time includes IO operations to convert the data in this case, but in the tests above (my computer) I used LightGBM in the command line with data already converted.

And htop:

![captura de tela 2016-11-23 as 20 36 24](https://cloud.githubusercontent.com/assets/9610024/20580911/8df42aca-b1bc-11e6-8404-8bb2fbf26286.png)


Of course it's faster because of more cores, but something in Sierra (or my computer specifically) doesn't run it with a bigger user-to-real time ratio.

Even running this way LightGBM is quite fast, so it has already improved my workflow.

**Comment by guolinke** [comment]
@ledmaster  It is very strange for you cost much "sys" time on OS X. It almost half of total time.
But from wxchan's log, it is much better. 
Could you give more details about your environment and compiler ? 
@chivee could you help to dig this?

**Comment by wxchan** [comment]
Actually I have the same issue.

I run examples/regression for 100k iterations, the time is:
real    16m15.995s
user    12m22.990s
sys     8m44.370s

and htop status is the same.

**Comment by guolinke** [comment]
@wxchan can you try to set num_threads=2 ?

**Comment by wxchan** [comment]
@guolinke 
num_threads=2:
real    13m36.647s
user    10m59.049s
sys     4m42.825s

**Comment by ledmaster** [comment]
@guolinke Sure, can you tell me what type of information would be useful?


For now:
My compiler is GCC/G++ 6.2.0.

My environment is:
macOS Sierra
Intel Core i7 2.8ghz
16gb 1333MHz DDR3

**Comment by guolinke** [comment]
@wxchan  Did you figure why osx cost some much time in ```sys``` ?

**Comment by wxchan** [comment]
Not yet. It looks proportional to the number of iterations. Do you have any suggestions to help find out the reasons?

**Comment by guolinke** [comment]
@wxchan 
in the latest commits, i update some omp loop strategy and gain some speed-up.
Can you test this in OS X and report the sys time ? Thanks

**Comment by ledmaster** [comment]
@guolinke While you wait for @wxchan results, this is running it on a 180k rows dataset:

real	4m12.452s
user	5m3.590s
sys	3m17.036s

If I run XGBoost with their new binning capabilities it doesn't seem to use all the threads in htop too.

**Comment by guolinke** [comment]
@ledmaster 
I find openmp cannot work well in some CPUs, recently. 
Can you provide your cpu info?

**Comment by ledmaster** [comment]
@guolinke 
Intel Core i7, 2.8ghz, 2 cores.

**Comment by guolinke** [comment]
@ledmaster if the physical cores are only 2, you better set num_threads=2.

**Comment by wxchan** [comment]
I forgot to reply here.

examples/regression for 100k iterations, num_threads=2:

real	15m43.995s
user	12m56.495s
sys	6m43.981s

real	16m20.021s
user	12m33.574s
sys	6m29.688s

**Comment by xelibrion** [comment]
Experienced the same issue today, it is actually running 4 threads just as I told it too, it's just seem to unable to utilise all cores. There's quite a bit of context switching happening as well - not sure if that's intended. 

<img width="588" alt="screen shot 2017-04-04 at 7 04 20 pm" src="https://cloud.githubusercontent.com/assets/175334/24649192/948e5ecc-1969-11e7-9293-4187d3699e92.png">

**Comment by Laurae2** [comment]
@xelibrion Just for debugging purposes: can you tell me what you get if you compile this?

```cpp
#include <omp.h>
#include <stdio.h>

int main() {
    #pragma omp parallel
    printf("nthread %d, thread %d\n", omp_get_num_threads(), omp_get_thread_num());
}
```

This should use all your cores to print (I suppose it will use all 4 threads).

@ledmaster xgboost fast histogram requires large dataset to be able to keep 100% CPU usage constant. Usually, it will never reach 100% for anything other than singlethread (unless your dataset is scaling largely enough to the number of threads).
